datacube/api/core.py

```diff
@@ -686,7 +686,7 @@ def group_datasets(
         datasets: Iterable[Dataset], group_by: GroupBy
     ) -> xarray.DataArray:
```

Comment:
The `group_by` parameter here should take the same type signature as `query_group_by` below.

(I noticed this was wrong yesterday - then when I saw this PR I thought "Yay!  Peter has fixed it already!!" but you didn't bubble the changes up to the top level API)

Reply:
The code in this function is clearly not written to handle being passed None, so I omitted that.

Looking at the other function, I don't think it makes sense to accept None in that either, so I removed it, and the type checker still passes. Considering the tests and integration tests are type checked, I think that's a reasonably strong statement about what parameters can be expected.